### ISSUE TITLE
feature: download-archive for subjectively unrecoverable errors (HttpErrors, video account terminated, etc)

### ISSUE THREAD

**Issue by chapmanjacobd** [yt-dlp/yt-dlp#3605]
### Checklist

- [X] I'm reporting a feature request
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

It would be useful to have a mechanism to remember and skip over (subjective) unrecoverable errors. I was in the middle of solving this with a `yt-dlp` wrapper script but the more I get into it I realize the problem is a little complex.

> Upon receiving a 410 status code, the client should not request the resource in the future 
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

¯\\_(ツ)_/¯ 

I've categorized the errors into three groups. 

- `X` Video not likely to ever be available
- `/` Video might be available in the future
- `O` Video seems available but not in the current geolocation

Status | DownloadArchive | ErrorArchive
-- | -- | --
HTTP 200 | 🍀 |  
HTTP 410/404 |   | X
HTTP 400/403/411 |   | /
HTTP 500/501/502/504 |   | /
Video unavailable |   | X
Video unavailable. This video is no longer available because the X account associated with this video has been terminated. |   | X
This video has been removed for violating X policy on harassment and bullying |   | X
This video doesn't exist |   | X
... has blocked it on copyright grounds ... |   | X
This video has been removed for violating X Terms of Service |   | X
This video is private |   | /
... no longer available due to a copyright claim |   | /
... is not currently available |   | /
... blocked it in your country |   | O
The uploader has not made this video available in your country |   | O
It is not available in your country |   | O

Remembering error codes `HTTP 1xx/3xx/418/420/429/431/503/509/507/511` are likely not very useful

There are a few options that I've thought up but I'm sure others are more creative:

1. Change the default behavior of download-archive to remember the most objectively invalid errors ("X") and add a flag (or two) to increase the **aggressiveness of remembering** (eg. a flag to enable remembering "/"-category errors and another flag for "O" [or just not implement the "O"-category flag--I only mention it to make sure that "O" category errors are not treated the same as "/"]).

2. Adding a new flag `--error-archive`. The file structure would be exactly the same as `--download-archive` such that--if they need to be different files due to implementation--one could concat the `error-archive` into the `download-archive` file. Being able to use the same file for both `--error-archive` and `--download-archive` is a nice-to-have but absolutely not required to be useful.

Luckily, this is not a huge issue but about 30% of the links that I encounter are invalid for one reason or another so it would help to save some time/bandwidth.

Issue https://github.com/yt-dlp/yt-dlp/issues/3515 is kind of similar. I couldn't find another issue more closely related

**Comment by pukkandan** [comment]
This feature was briefly discussed here https://github.com/yt-dlp/yt-dlp/issues/689#issuecomment-899101608. While there is no other issue open for this exact request, https://github.com/yt-dlp/yt-dlp/issues/1761 is a subset of this (Making https://github.com/yt-dlp/yt-dlp/issues/1659, https://github.com/yt-dlp/yt-dlp/discussions/876, #457 related issues). Also see [this reddit post](https://www.reddit.com/r/youtubedl/comments/omcw10/log_file_for_skippedbad_downloads)

There are quite a few issues in implementing this (Keep in mind youtube is not the only supported site):
1. There is some implementation difficulty in propagating the error information to the region of code that handles archives
2. We may not know the video ID if it errors. For youtube, the ID is always in the URL, so we can always determine it. But this is not true for other sites
3. yt-dlp does not know the difference b/w recoverable and irrecoverable errors.
4. Actually, the distinction b/w these is not clear even for a human. Many of the errors you have marked as irrecoverable are actually recoverable in certain edge cases. Eg: mistakenly deleted videos may be re-instated by youtube. Or more likely example, videos blocked in a region gets unblocked all the time. Similarly, a 404 could just indicate that the website is down (or even that the extractor is broken), not necessarily that the video doesnt exist, etc

Considering all these, I believe the closest to this we can ever implement is #1761 (even which is pretty difficult)

**Comment by chapmanjacobd** [comment]
> We may not know the video ID if it errors

ohhh that is a really good point

I refactored some of my code and found that it wasn't too difficult to get something working. If people have a similar issue as me in the future they can look at this for inspiration. My big _a-ha_ moment was realizing that completely separating the link-gathering code from the downloading code makes it a lot easier to restart, debug, and keep track of status and links

https://gist.github.com/chapmanjacobd/a755735b3437cd9a8489d826a3d1715d

Feel free to close this issue whenever. I'm going to leave it open for a bit so more people get some ideas if they have the same problem

**Comment by chapmanjacobd** [comment]
Closing in favor of https://github.com/yt-dlp/yt-dlp/issues/457
